pkg/auditlog/publisher.go

```diff
@@ -0,0 +1,158 @@
+package alog
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+// Publisher represents an audit log events publisher that posts events to
+// WorkOS asynchronously.
+type Publisher struct {
+	// The WorkOS api key. It can be found in
+	// https://dashboard.workos.com/api-keys.
+	APIKey string
+
+	// The http.Client that is used to post audit log events to WorkOS. Defaults
+	// to http.Client.
+	Client *http.Client
+
+	// The endpoint used to request Workos. Defaults to
+	// https://api.workos.com/events.
+	Endpoint string
+
+	// The function used to encode in JSON. Defaults to json.Marshal.
+	JSONEncode func(v interface{}) ([]byte, error)
+
+	// The function used to log errors. Defaults to log.Printf.
+	Log func(format string, v ...interface{})
+
+	// The size of the internal queue. Defaults to 512.
+	QueueSize int
+
+	queue  chan Event
+	cancel func()
+	stop   chan struct{}
+	once   sync.Once
+}
+
+// Publish enqueues the given events to be published to WorkOS.
+func (p *Publisher) Publish(events ...Event) {
+	p.once.Do(p.init)
+
+	for _, e := range events {
+		e.Location = defaultLocation(e.Location)
+		e.OccurredAt = defaultTime(e.OccurredAt)
+		p.queue <- e
+	}
+}
+
+func (p *Publisher) init() {
+	if p.Client == nil {
+		p.Client = &http.Client{Timeout: 10 * time.Second}
+	}
+
+	if p.Endpoint == "" {
+		p.Endpoint = "https://api.workos.com/events"
+	}
+
+	if p.JSONEncode == nil {
+		p.JSONEncode = json.Marshal
+	}
+
+	if p.Log == nil {
+		p.Log = log.Printf
+	}
+
+	if p.QueueSize < 1 {
+		p.QueueSize = 512
+	}
+
+	p.queue = make(chan Event, p.QueueSize)
+	p.stop = make(chan struct{})
+
+	ctx := context.Background()
+	ctx, p.cancel = context.WithCancel(ctx)
+
+	go p.loop(ctx)
+}
+
+func (p *Publisher) loop(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			p.stop <- struct{}{}
+			return
+
+		case e := <-p.queue:
+			// This is to capture e value in order to not have the same value
+			// passed in different goroutines.
+			event := e
+			event.indempotencyKey = uuid.New().String()
```

Comment:
It seems like the idempotency key should be set when the event is created, otherwise if they have a reference to an event struct they can pass the same event twice.

https://github.com/workos-inc/workos-go/blob/8ca00a6ef4cda750c32cf86075932a89c1afcac8/auditlog/event.go#L78-L92

Reply:
The event is passed by value. Event if they keep a reference to the event they published,
The event that is passed to the channel is a copy of the one they used in the `Publish` call. The reference they have will always have an indempotency key empty.

Reply:
> The event that is passed to the channel is a copy of the one they used in the `Publish` call.

Since the idempotency key isn't set until the channel receives it in this implementation, if the client sends the same struct again a new idempotency key will be generated for the same event.

Reply:
> It feels like reusing a same event is a developer mistake. Each event should be unique whether they embeds the same data.

For sure! I see the benefits of both implementations. I think clients should help guard against developer mistakes where it can. It will boil down to the semantics the client wants to enforce here, is the client treating the requests as idempotent or the events? If events are idempotent, the requests will be as a result. Just my thoughts on it.

Reply:
Since it is documented in the the API, is the idempotency key should be a choice for the user?
We could just make this field exported and fill it only if it is not set by the user.

What I'm trying to keep is a declaration by structure since each field document by itself the Event content.

I would also like to keep passing an event by value in order to prevent a memory allocation on the stack.